[Tablet Orders] Persist in-flight order details on size class changes

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12252

## Description
This PR addresses the issue of modal views being dismissed automatically on size class changes, which produces a loss of the data the merchant may already have entered, but not specifically saved.

We tackle the cases for `Customer Notes` and `Customer Details`, these are the ones where a customer might have written more data manually into the fields and prone to be lost upon accidental dismissal.

**Question:** I've been experimenting with applying the same behaviour for Custom Amounts, however, I'm not convinced is the right way to go and perhaps we should just allow the custom amount to be lost upon dismissal? If we do, I believe we can tackle it separately, since we may need to rethink the current implementation of custom amounts as it relies on being re-created and reset multiple times. The work-in-progress experimental branch can be found here: `experiment/12252-continuation-persist-custom-amount-on-size-class-change`

## Changes
- When we initialize an `EditableOrderViewModel`, we'll observe and keep a reference to the latest changes to the customer details object. If there's a size class change triggered from the `OrderForm` then we'll save this latest value to the order.

- With order notes, we just perform a note update when the size class change is triggered.

## Testing instructions
1. Launch the app on an iPad
2. Open the Orders tab and tap `+`
3. Tap `Add Customer Details`
4. Enter some text
5. Change the size class using a split view so that only one pane is shown
6. Observe that the modal view has been dismissed automatically, but the data has persisted in the order
7. Repeat the process for `Add Note`

## Screenshots

![Simulator Screen Recording - iPad Pro (12 9-inch) (6th generation) - 2024-03-20 at 10 02 35](https://github.com/woocommerce/woocommerce-ios/assets/3812076/5e5663c1-6ff9-41ec-a349-4077d0915a60)

iamgabrielma committed:
Save in-flight changes to customer details

iamgabrielma committed:
Save in-flight changes to customer notes

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Add unit tests

iamgabrielma commented:
@woocommerce/mobile-ui-testing-squad I would need some help with the UI tests in this PR: 

I can see that the UI test for `OrdersTests.test_load_existing_order()` on iPhone has failed twice in CI, but I'm unable to run the UI test suite locally to confirm where the problem might be.

When I try locally (with the WooCommerceUITests target, and WireMock running), Xcode never finishes the building process. We only see the testing suite has completed (successfully, because no tests were run) when I stop it manually:

<img width="926" alt="Screenshot 2024-03-20 at 10 08 54" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/188e9e66-71aa-4443-9ecd-f1e3832e858e">

I'll continue debugging my dev setup, but for the time being would be great to get some input regarding the failing test, the only information we got is a timeout:

```
🔬 UI Tests (iPhone): 1 tests have failed (1 distinct assertion failure(s) in total)
<details><summary><tt>test_load_existing_order()</tt> in <tt>OrdersTests</tt> (2 times)</summary>
<pre>failed: caught error: "timedOut"</pre>
```

jostnes commented:
@iamgabrielma i haven't run WCiOS repo/UI tests in a while so it might take me some time to do that, but i checked the `xcresult.zip` file from the failing build and noticed this (the cool thing about the xcresult file is that it now records the test in action so you can follow along and see where it went wrong):
<img width="1127" alt="Screenshot 2024-03-20 at 4 15 44 PM" src="https://github.com/woocommerce/woocommerce-ios/assets/17252150/c47c67c4-cec6-4d58-8724-4014d700c448">

from there, it looks like after tapping the `Done` button, an error is displayed, and the save fails. if this flow works without issue during manual testing, my main suspicion is that something in the response that the mock file was expecting changed and needs to be updated (theory not tested)

(also, looks like the error message is truncated, should that be fixed?)

jostnes commented:
I managed to get the tests running locally and added some breakpoints to the test to confirm the theory. 

Using this PR branch, when tapping on the Edit button, WireMock expects to get a response that is missing:

https://github.com/woocommerce/woocommerce-ios/assets/17252150/9fbf7ab8-a550-4c2d-b551-a44f84848adc

The same test on trunk (where the test is passing), doesn't happen:

https://github.com/woocommerce/woocommerce-ios/assets/17252150/140f55f9-d69b-4af1-baa8-438aa5c344e7

There are differences between this branch and trunk, but I'm unsure if the case is that the test is missing a mock file or if it's a bug since it's happening on iPhone but not iPad. Is this expected behavior from the change in this PR?

joshheald commented:
> There are differences between this branch and trunk, but I'm unsure if the case is that the test is missing a mock file or if it's a bug since it's happening on iPhone but not iPad. Is this expected behavior from the change in this PR?

Thanks @jostnes 

It looks like this is is a minor bug, as we're syncing the customer details and the order notes when we open the order for editing – those are the unexpected requests.

This happens on both iPhone and iPad, but on iPhone the resulting notice blocks the done button, which is why the test is failing.

I'm working on a code change to prevent the call when we first load the screen, as it doesn't really make sense to sync something we just loaded, there's no changes to save.

joshheald committed:
12252 Prevent extra sync on open order creation

When we opened the create/edit order screen, we attempted to sync the order notes and customer details.

This was only supposed to happen on size class change – to avoid losing user entered data.

This commit changes it to only save when there’s a change.

We could still improve this by making the sync happen only when there’s a change from the synced value.

joshheald committed:
Merge branch 'trunk' into issue/12252-persist-data-on-size-class-change

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -743,6 +757,25 @@ final class EditableOrderViewModel: ObservableObject {
+    /// Saves the latest Customer Details state if the view is dismissed. Eg: on IPads, automatic modal dismissal on size class change
+    ///
+    func saveInflightCustomerDetails() {
+        let address = Address(firstName: latestAddressFormFields?.firstName ?? "",
+                              lastName: latestAddressFormFields?.lastName ?? "",
+                              company: latestAddressFormFields?.company,
+                              address1: latestAddressFormFields?.address1 ?? "",
+                              address2: latestAddressFormFields?.address2,
+                              city: latestAddressFormFields?.city ?? "",
+                              state: latestAddressFormFields?.state ?? "",
+                              postcode: latestAddressFormFields?.postcode ?? "",
+                              country: latestAddressFormFields?.country ?? "",
+                              phone: latestAddressFormFields?.phone,
+                              email: latestAddressFormFields?.email)
```
```suggestion
        guard let latestAddressFormFields else {
            return
        }
        let address = Address(firstName: latestAddressFormFields.firstName,
                              lastName: latestAddressFormFields.lastName,
                              company: latestAddressFormFields.company,
                              address1: latestAddressFormFields.address1,
                              address2: latestAddressFormFields.address2,
                              city: latestAddressFormFields.city,
                              state: latestAddressFormFields.state,
                              postcode: latestAddressFormFields.postcode,
                              country: latestAddressFormFields.country,
                              phone: latestAddressFormFields.phone,
                              email: latestAddressFormFields.email)
```

This might be a _start_ for preventing unnecessary syncs, i.e. when the address hasn't changed, we don't do anything.

It won't fix that fully, I don't think, because changes which have been made and synced would get synced again...

iamgabrielma replied:
Sounds good, I've added your changes along with a check to the latest order synchronizer state on https://github.com/woocommerce/woocommerce-ios/pull/12306/commits/20ef604e2c024e7fd800af27c85eec3887857707

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -789,6 +822,12 @@ final class EditableOrderViewModel: ObservableObject {
+    /// Saves the current contents of the Order Note
+    ///
+    func saveInFlightOrderNotes() {
+        updateCustomerNote()
```
We need some gate on this, we shouldn't sync the order notes (and track that they've changed) if there's no changes or we've previously synced them. Especially not repeatedly in response to rotation...

iamgabrielma replied:
True! I've updated it to use the same approach as with customer details, through checking the latest state of the note within the order synchronizer: https://github.com/woocommerce/woocommerce-ios/pull/12306/commits/a5b491a6038ef158e62c83908b9efbabb53340af

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/OrderForm.swift:
```diff
@@ -148,6 +148,11 @@ struct OrderFormPresentationWrapper: View {
                     .accessibilityIdentifier(OrderForm.Accessibility.cancelButtonIdentifier)
                 },
                 isShowingSecondaryView: $viewModel.isProductSelectorPresented)
+            .onChange(of: horizontalSizeClass) { _ in
+                // Save: in-flight order details when there's size class changes
+                viewModel.saveInflightCustomerDetails()
+                viewModel.saveInFlightOrderNotes()
+            }
```
I moved this here, to fix the UI test – this avoids us calling these methods when the view first appears.

iamgabrielma replied:
This update wasn't working on my end, I'm not entirely sure why, but the `.onChange` was never called upon size class changes so the methods weren't called either.

I suspect it could be due to the ViewBuilder redrawing the whole thing, which wouldn't trigger the .onChange? After some experimentation it does work properly if we add it here instead: https://github.com/woocommerce/woocommerce-ios/pull/12306/commits/8a9bbc97bf29e8963775db3cd9f345890b22c094

joshheald replied:
Hmm.

That seems plausible, except it must have been working for me, or I wouldn't have put the code there. Very strange.

Where you've moved it, you don't need to use the `horizontalSizeClass`, as you have `presentationStyle` already.

iamgabrielma replied:
> Where you've moved it, you don't need to use the horizontalSizeClass, as you have presentationStyle already.

That was my original approach, but it also happens here that if we switch the `.onChange(of: horizontalSizeClass)` for `.onChange(of: presentationStyle)`, the `onChange()`  is never called when there's an update, and the in-flight data not saved. 

I've wonder if could be related to the usage of `@Environment` for the custom presentationStyle value, and if the behaviour is slightly different than other wrappers when updating/publishing values. I found comments regarding different behaviour and the need to use view modifiers like `.environment(\.key, value)` or changes might not propagate as expected, but at this point everything got a bit hairy to debug.

joshheald reviewed (commented):
I think this needs some more work – I've fixed the UI test, but the underlying issue is that we're syncing address and notes even when they haven't changed, and tracking that they've been changed.

A couple of suggestions on how to do it inline... essentially we need to check whether the synced state matches the UI, and only sync if it doesn't. That would match the product selection behaviour, at least.

Even better would be to save the entered changes locally, and repopulate them in the UI after the size class change... I had a poke around at changing the AdaptiveModalContainer to avoid redrawing the views totally, but couldn't do that quickly. I think that's the root cause of some of the problems here though.

iamgabrielma committed:
Merge branch 'trunk' into issue/12252-persist-data-on-size-class-change

# Conflicts:
#	WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/EditableOrderViewModelTests.swift

iamgabrielma committed:
Merge branch 'issue/12252-persist-data-on-size-class-change' of https://github.com/woocommerce/woocommerce-ios into issue/12252-persist-data-on-size-class-change

dangermattic commented:
<!--
  0 Errors
  1 Warning: PR is not assigned to a milest...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
update note only when differs from the ordersync

Updates the in-flight note only when this is different with what is already saved in the order note

iamgabrielma committed:
Hook observation again after form reset

Observing published values for customer details only worked the first time we edit the form, but not after.

The issue comes from the form being deallocated upon dismissal. As we recreate the form on editing, at this point we’ve lost the reference that was set on EditableOrderViewModel init by `observeChangesInCustomerDetails()`, hence not observing new values that are published through the new object

This is fixed by calling again the observation hook when the reference is updated.

iamgabrielma committed:
Save in-flight customer details only when there’s diff

iamgabrielma committed:
save in-flight data on orderFormSummary size class change

iamgabrielma committed:
Merge branch 'trunk' into issue/12252-persist-data-on-size-class-change

# Conflicts:
#	WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/EditableOrderViewModelTests.swift

iamgabrielma commented:
Thanks for the tips and debugging the UITest @jostnes ! Learned something new about UITests and the zip file results :D. I'm still having troubles running them locally, but seems to be entirely related to my local setup and Xcode version.

@joshheald  When you have the time, this is ready for review.

I found a new issue with in-flight customer details not being updated after the first time its updated, which was related to the  form being deallocated and recreated on dismissal, so we were losing the publisher reference that we create when we observe changes on initialization, and new values were just not observed anymore. This is fixed on https://github.com/woocommerce/woocommerce-ios/pull/12306/commits/d3b67b89b2926b67b8b5d197816eff7bb1ae4515

Now in-flight changes to order notes and customer details are only saved when there's an actual difference from what's already stored in the latest order sync.

joshheald reviewed (approved):
Works as expected, thanks for seeing this one through.

We need to be really mindful of this kind of thing in future... because while we've protected against some data loss, saving without warning on rotation isn't a great way to solve the problem.

This approach could also (less easily) cause data loss, e.g. a user opens the note, accidentally clears it, then rotates the device. There's no undo or way back, except maybe via the web... it's just gone. 

I still think this is better than nothing, but avoiding the redraw would be significantly better still... I just don't know how to do that in a sane way yet.
